src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -594,6 +597,43 @@ protected RubyHash recursiveObjects(
 
     }
 
+    @Primitive(name = "thread_detect_recursion_single")
+    public static abstract class ThreadDetectRecursionSingleNode extends PrimitiveArrayArgumentsNode {
+        @Child private YieldNode dispatchNode = YieldNode.create();
+        @Child private SetNode addNode = SetNode.create();
+        @Child private DeleteLastNode delNode = DeleteLastNode.create();
+
+
+        public Object yield(RubyProc block, Object... arguments) {
+            return dispatchNode.executeDispatch(block, arguments);
+        }
+
+        protected boolean add(RubyHash hash, Object key, Object value) {
+            return addNode.executeSet(hash, key, value, true);
+        }
+
+        protected Object remove_last(RubyHash hash, Object key) {
+            return delNode.executeDeleteLast(hash, key);
+        }
+
+        @Specialization
+        protected boolean detectRecursionSingle(Object obj, RubyProc block,
+                @Cached GetCurrentRubyThreadNode getCurrentRubyThreadNode,
+                @Cached("createBinaryProfile()") ConditionProfile containsProfile) {
+            RubyHash objects = getCurrentRubyThreadNode.execute().recursiveObjectsSingle;
+            if (containsProfile.profile(add(objects, obj, true))) {
```

Comment:
`containsProfile` is confusing here, add() return true if added a new entry.
So I'd suggest `addedProfile`/`insertedProfile`.